File: ios/Roana/Roana/Motion/MotionQuality.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 The Roana Authors.

import Foundation

enum MotionQualityLabel: String {
    case stable
    case pointingDown = "pointing_down"
    case unstable
}

struct MotionQuality: Equatable {
    let label: MotionQualityLabel
    let reason: String

    var trustsGuidance: Bool {
        label == .stable
    }
}

struct MotionQualitySample: Equatable {
    let pitchDegrees: Double
    let angularVelocityDegreesPerSecond: Double

    init(pitchDegrees: Double, angularVelocityDegreesPerSecond: Double) {
        self.pitchDegrees = pitchDegrees
        self.angularVelocityDegreesPerSecond = angularVelocityDegreesPerSecond
    }
}

enum MotionQualityClassifier {
    static let pointingDownPitchDegrees = -55.0
    static let unstableAngularVelocityDegreesPerSecond = 120.0

    static func classify(_ sample: MotionQualitySample?) -> MotionQuality {
        guard let sample else {
            return MotionQuality(label: .stable, reason: "motion_unavailable")
        }

        if sample.pitchDegrees <= pointingDownPitchDegrees {
            return MotionQuality(label: .pointingDown, reason: "pitch_down")
        }

        if abs(sample.angularVelocityDegreesPerSecond) >= unstableAngularVelocityDegreesPerSecond {
            return MotionQuality(label: .unstable, reason: "high_angular_velocity")
        }

        return MotionQuality(label: .stable, reason: "motion_stable")
    }
}

File: app/src/main/java/com/roana/app/MotionQuality.kt
package com.roana.app

data class MotionQuality(
    val label: Label,
    val reason: String,
) {
    val trustsGuidance: Boolean
        get() = label == Label.STABLE

    enum class Label(
        val logValue: String,
    ) {
        STABLE("stable"),
        POINTING_DOWN("pointing_down"),
        UNSTABLE("unstable"),
    }
}

data class MotionQualitySample(
    val pitchDegrees: Double,
    val angularVelocityDegreesPerSecond: Double,
)

object MotionQualityClassifier {
    const val POINTING_DOWN_PITCH_DEGREES = -55.0
    const val UNSTABLE_ANGULAR_VELOCITY_DEGREES_PER_SECOND = 120.0

    fun classify(sample: MotionQualitySample?): MotionQuality {
        if (sample == null) {
            return MotionQuality(MotionQuality.Label.STABLE, "motion_unavailable")
        }

        if (sample.pitchDegrees <= POINTING_DOWN_PITCH_DEGREES) {
            return MotionQuality(MotionQuality.Label.POINTING_DOWN, "pitch_down")
        }

        if (kotlin.math.abs(sample.angularVelocityDegreesPerSecond) >=
            UNSTABLE_ANGULAR_VELOCITY_DEGREES_PER_SECOND
        ) {
            return MotionQuality(MotionQuality.Label.UNSTABLE, "high_angular_velocity")
        }

        return MotionQuality(MotionQuality.Label.STABLE, "motion_stable")
    }
}
